### PR TITLE
Some improvements around identity.

### PR DESCRIPTION
**Pull request by robertpustulka**
Refs #132

### CODE REVIEW

**Comment by robertpustulka** [comment]
No checks from travis?

**Comment by ADmad** [comment]
Strange.. I pushed a commit to master and travis build was triggered.

**Comment by ADmad** [comment]
The good 'ole close and reopen PR trick!

**Comment by robertpustulka** [comment]
Force push also retriggered the build. But shouldn't there be 2 of them?

**Comment by ADmad** [comment]
Nope, you made the `identity` branch on your fork.

**Comment by robertpustulka** [comment]
I found another bug, where the AuthenticationMiddleware would set the identity from a result (raw data) not the IdentityInterface object.

Should I fix it here or open another PR?

**Comment by robertpustulka** [comment]
I discovered that the identity object and identity data concepts were mixed up around the code.
That required updating tests to assure the `IdentityInterface` instances are attached to the request attribute.

`ResultInterface::getIdentity()` actually returns the identity data retrieved from identifiers, not the `IdentityInterface` object. The object is created by the `AuthenticationService`. The middleware attached the raw data to the request not the actual identity object. In order to fix this I had to expose `AuthenticationServiceInterface::getIdentifier()` method and use that to attach the request attribute.


I believe the name of method `ResultInterface::getIdentity()` is confusing. It does not get the identity, it gets the identity data. Should we rename the method to `getData` or `getIdentityData` to make the distinction clearer?

**Comment by markstory** [comment]
>  Should we rename the method to getData or getIdentityData to make the distinction clearer?

Good find on that inconsistency. `getData()` make sense to me.

File: src/Identity.php [hunk 1]
```diff
@@ -36,18 +38,25 @@ class Identity implements IdentityInterface
...
-     * @param array $identityData Identity data
+     * @param array|\ArrayAccess $identityData Identity data
      * @param array $config Config options
+     * @throws InvalidArgumentException When invalid identity data is passed.
      */
     public function __construct($identityData, array $config = [])
     {
+        if (!is_array($identityData) && !$identityData instanceof ArrayAccess) {
+            $type = is_object($identityData) ? get_class($identityData) : gettype($identityData);
+            $message = sprintf('Array data must be an `array` or implement `ArrayAccess` interface, `%s` given.', $type);
```
Comment (markstory) [inline src/Identity.php]: Should this be `identityData must be an array...`

**Comment by robertpustulka** [comment]
Identity data can only be an instance of `ArrayAccess` so there's no need to support arrays.

`Identity::toArray()` has been refactored to `getOriginalData()`. It's not easy to convert every implementation of `ArrayAccess` into an array.
Checking for existance of `toArray()` method and typecasting otherwise seems like a bad idea. I decided that having a getter for the original data object is enough.

File: src/Identity.php [hunk 2]
```diff
@@ -134,11 +135,11 @@ public function offsetUnset($offset)
...
 
     /**
-     * Turns the object into an array
+     * Gets the original data object.
      *
-     * @return array
+     * @return \ArrayAccess
      */
-    public function toArray()
+    public function getOriginalData()
```
Comment (markstory) [inline src/Identity.php]: :+1: to this. It is a fair bit safer to use.

**Comment by ADmad** [comment]
This makes me think what's the benefit of having identity data be an `ArrayAccess` instance?

Included identifiers return Cake' `Entity` instance so we can just use their properties directly and for external providers too you are rarely going to get objects which implement array access. It's lot more likely to get simple object of type stdclass with properties. E.g [1](https://github.com/SocialConnect/common/blob/master/src/Entity/User.php), [2](https://github.com/hybridauth/hybridauth/blob/master/src/User/Profile.php).

**Comment by robertpustulka** [comment]
We force the identity data to be `ArrayAccess` in order to easily retrieve the properties in `Identity` class, including the property for `getIdentifier()` method. You're right that this requirement may over complicate things in certain scenarios (for example when working with external libs). Someone might need to write a wrapper implementing the `ArrayAccess` interface for their internal identity classes so the `Identity` class can access the `id` property. The shorter way would be to return the `IdentityInterface` instance from the custom resolver, but still there's a requirement of `ArrayAccess`.

I think we could drop the requirement of `ArrayAccess` from `IdentityInterface` and expose a `getOriginal()` method within this interface. That may simplify things. 

We would need to remove any `ArrayAcces` type hints from where the identity data is concerned. The return data for `IdentifierInterface::identify()` and `ResultInterface::getData()` would need to be `mixed` as well.

The default `Identity` class could still require an `ArrayAccess` in it's constructor to easily blend in with CakePHP ORM. Custom resolvers could return `IdentifierInstance` so the default `Identity` class does not get involved. Or the developer could provide their own identity class to the authentication service.

**Comment by markstory** [comment]
Having `IdentityInterface` extend `ArrayAccess` lets us provide more features to users of the plugin without being an onerous or difficult to implement interface.

**Comment by robertpustulka** [comment]
@markstory OK right. So how about identity data and the requirement of `ArrayAccess`? Should we be a bit more liberal about it and allow regular arrays as well? We could allow any kind of data as well, but I'm not sure how  the data would be converted to the `Identity` in that case...

**Comment by markstory** [comment]
>  So how about identity data and the requirement of ArrayAccess? Should we be a bit more liberal about it and allow regular arrays as well?

I think the identity data being an array or ArrayAccess will be fine.

File: src/Identity.php [hunk 3]
```diff
@@ -37,18 +38,25 @@ class Identity implements IdentityInterface
...
-     * @param array $config Config options.
+     * @param array|\ArrayAccess $data Identity data
+     * @param array $config Config options
+     * @throws InvalidArgumentException When invalid identity data is passed.
      */
-    public function __construct(ArrayAccess $data, array $config = [])
+    public function __construct($data, array $config = [])
     {
+        if (!is_array($data) && !$data instanceof ArrayAccess) {
+            $type = is_object($data) ? get_class($data) : gettype($data);
```
Comment (ADmad) [inline src/Identity.php]: We should have a function in core itself for this :)

**Comment by robertpustulka** [comment]
Is there anything to be done here?

**Pull request merged**